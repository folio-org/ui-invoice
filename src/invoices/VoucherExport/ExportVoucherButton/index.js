export { default as ExportVoucherButton } from './ExportVoucherButton';
