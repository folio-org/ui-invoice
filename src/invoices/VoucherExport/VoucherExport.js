import {
  useCallback,
  useEffect,
  useMemo,
  useState,
} from 'react';
import { useLocation, useHistory } from 'react-router';
import { FormattedMessage, useIntl } from 'react-intl';
import { noop } from 'lodash';

import {
  checkScope,
  Col,
  HasCommand,
  Loading,
  Pane,
  Paneset,
  Row,
  Select,
} from '@folio/stripes/components';
import {
  handleKeyCommand,
  RESULT_COUNT_INCREMENT,
  useSorting,
} from '@folio/stripes-acq-components';

import { getBatchGroupsOptions } from '../../common/utils';
import { BatchVoucherExportsList } from './BatchVoucherExportsList';
import { BV_EXPORT_SORTABLE_COLUMNS } from './BatchVoucherExportsList/constants';
import { ConfirmManualExportModal } from './ConfirmManualExportModal';
import { VoucherExportFooter } from './VoucherExportFooter';
import {
  useBatchGroupExportConfigs,
  useBatchGroups,
  useBatchVoucherExports,
  useManualExportRun,
} from './hooks';

const VoucherExport = () => {
  const history = useHistory();
  const location = useLocation();
  const intl = useIntl();

  const [selectedBatchGroupId, setSelectedBatchGroupId] = useState();
  const [confirmManualExportModalOpen, setConfirmManualExportModalOpen] = useState(false);
  const [pagination, setPagination] = useState({ limit: RESULT_COUNT_INCREMENT, offset: 0 });
  const [
    sortingField,
    sortingDirection,
    changeSorting,
  ] = useSorting(noop, BV_EXPORT_SORTABLE_COLUMNS);

  const {
    batchGroups,
    isLoading: isBatchGroupsLoading,
  } = useBatchGroups();
  const {
    batchVoucherExports,
    totalRecords,
    isFetching,
    isLoading: isBatchVoucherExportsLoading,
    refetch,
  } = useBatchVoucherExports(
    selectedBatchGroupId,
    pagination,
    { sortingField, sortingDirection },
  );
  const {
    exportConfigs,
    isLoading: isExportConfigsLoading,
  } = useBatchGroupExportConfigs(selectedBatchGroupId);

  const [runManualExport, clearManualExportRun] = useManualExportRun({
    batchGroupId: selectedBatchGroupId,
    batchGroups,
    batchVoucherExports,
    refetch,
  });

  useEffect(() => {
    return () => {
      clearManualExportRun();
    };
  }, []);

  useEffect(() => {
    if (batchGroups.length === 1) setSelectedBatchGroupId(batchGroups[0].id);
  }, [batchGroups]);

  const isLoading = (
    isBatchGroupsLoading
    || isBatchVoucherExportsLoading
    || isExportConfigsLoading
  );

  const batchGroupsOptions = useMemo(() => (
    batchGroups.length > 1
      ? [
        {
          label: intl.formatMessage({ id: 'ui-invoice.voucherExport.batchGroup.select' }),
          value: '',
        },
        ...getBatchGroupsOptions(batchGroups),
      ]
      : getBatchGroupsOptions(batchGroups)
  ), [batchGroups, intl]);

  const selectedBatchGroupName = useMemo(() => (
    batchGroups.find(({ id }) => id === selectedBatchGroupId)?.name
  ), [batchGroups, selectedBatchGroupId]);

  const onCancel = useCallback(() => (
    history.push({
      pathname: '/invoice',
      search: location.search,
    })
  ), [history, location.search]);

  const applySorting = (e, meta) => {
    changeSorting(e, meta);
    setPagination(prev => ({ ...prev, offset: 0 }));
  };

  const shortcuts = [
    {
      name: 'cancel',
      shortcut: 'esc',
      handler: handleKeyCommand(onCancel),
    },
    {
      name: 'search',
      handler: handleKeyCommand(() => history.push('/invoice')),
    },
  ];

  const isManualExportDisabled = isLoading || !selectedBatchGroupId || !exportConfigs?.format;

  const paneFooter = (
    <VoucherExportFooter
      disabled={isManualExportDisabled}
      onCancel={onCancel}
      runManualExport={() => setConfirmManualExportModalOpen(true)}
    />
  );

  return (
    <HasCommand
      commands={shortcuts}
      isWithinScope={checkScope}
      scope={document.body}
    >
      <Paneset>
        <Pane
          defaultWidth="100%"
          dismissible
          id="pane-voucher-export"
          onClose={onCancel}
          paneTitle={<FormattedMessage id="ui-invoice.voucherExport.paneTitle" />}
          footer={paneFooter}
        >
          <Row>
            <Col xs={4}>
              <Select
                disabled={isLoading}
                dataOptions={batchGroupsOptions}
                label={<FormattedMessage id="ui-invoice.settings.batchGroupConfiguration.batchGroup" />}
                onChange={({ target }) => setSelectedBatchGroupId(target.value)}
                fullWidth
                required
              />
            </Col>
          </Row>
          {
            selectedBatchGroupId && (
              isLoading
                ? <Loading />
                : (
                  <BatchVoucherExportsList
                    columnIdPrefix="voucher-export"
                    batchVoucherExports={batchVoucherExports}
                    format={exportConfigs?.format}
                    isLoading={isFetching}
                    onNeedMoreData={setPagination}
                    pagination={pagination}
                    sortDirection={sortingDirection}
                    onHeaderClick={applySorting}
                    totalCount={totalRecords}
                  />
                )
            )
          }

          <ConfirmManualExportModal
            open={confirmManualExportModalOpen}
            onConfirm={() => {
              runManualExport();
              setConfirmManualExportModalOpen(false);
            }}
            onCancel={() => setConfirmManualExportModalOpen(false)}
            selectedBatchGroupName={selectedBatchGroupName}
          />

        </Pane>
      </Paneset>
    </HasCommand>
  );
};

export default VoucherExport;
