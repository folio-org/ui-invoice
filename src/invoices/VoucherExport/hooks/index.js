export { useBatchGroups } from './useBatchGroups';
export { useBatchGroupExportConfigs } from './useBatchGroupExportConfigs';
export { useBatchVoucherExports } from './useBatchVoucherExports';
export { useManualExportRun } from './useManualExportRun';
