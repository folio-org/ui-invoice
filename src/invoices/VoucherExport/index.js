export { default as VoucherExport } from './VoucherExport';
