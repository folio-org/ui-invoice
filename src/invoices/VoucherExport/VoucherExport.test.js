import React from 'react';
import { render } from '@testing-library/react';

import VoucherExport from './VoucherExport';

jest.mock('react-router', () => ({
  ...jest.requireActual('react-router'),
  useHistory: jest.fn(),
  useLocation: jest.fn().mockReturnValue({}),
}));
jest.mock('@folio/stripes-components/lib/Commander', () => ({
  HasCommand: jest.fn(({ children }) => <div>{children}</div>),
}));

const renderVoucherExport = () => render(
  <VoucherExport />,
);

describe('VoucherExport', () => {
  beforeEach(() => {
    global.document.createRange = global.document.originalCreateRange;
  });

  afterEach(() => {
    global.document.createRange = global.document.mockCreateRange;
  });

  it('should render correct structure for voucher export', () => {
    const { asFragment } = renderVoucherExport();

    expect(asFragment()).toMatchSnapshot();
  });
});
