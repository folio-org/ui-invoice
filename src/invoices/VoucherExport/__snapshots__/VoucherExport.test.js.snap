// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VoucherExport should render correct structure for voucher export 1`] = `
<DocumentFragment>
  <div>
    <div
      class="paneset"
    >
      <section
        class="pane focusIndicator"
        id="pane-voucher-export"
        style="flex: 0 0 100%;"
      >
        <div
          aria-labelledby="paneHeaderpane-voucher-export-pane-title"
          class="paneHeader"
          data-test-pane-header="true"
          id="paneHeaderpane-voucher-export"
          role="region"
          tabindex="-1"
        >
          <div
            class="paneHeaderButtonsArea first"
          >
            <div
              class="paneMenu"
            >
              <button
                aria-label="stripes-components.closeItem"
                class="iconButton medium paneHeaderCloseIcon PaneHeaderIconButton paneHeaderIconButton"
                data-test-pane-header-dismiss-button="true"
                icon="times"
                type="button"
              >
                <span
                  class="iconButtonInner mediumOffset PaneHeaderIconButton__inner"
                  data-test-pane-header-dismiss-button="true"
                >
                  <span>
                    Icon
                  </span>
                </span>
              </button>
            </div>
          </div>
          <div
            class="paneHeaderCenter"
          >
            <div
              class="paneHeaderCenterInner"
              id="paneHeaderpane-voucher-export-pane-title"
            >
              <h2
                class="paneTitle"
                data-test-pane-header-title="true"
              >
                <span
                  class="paneTitleLabel"
                >
                  ui-invoice.voucherExport.paneTitle
                </span>
              </h2>
            </div>
          </div>
        </div>
        <div
          class="paneContent hasPadding"
          id="pane-voucher-export-content"
          tabindex="-1"
        />
      </section>
    </div>
    <div
      class="container"
    >
      <div
        class="cursorHover"
        hidden=""
        style="left: -10px;"
      >
        <div
          class="cursor"
        >
          Â 
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
